### DISCUSSION TITLE
ARM architecture support

### DISCUSSION THREAD

**Discussion by denysdovhan** [plausible/analytics#903] [Self-Hosted Support] [answered]
First of all, thanks for your awesome project!

I'd love to host plausible on my Raspberry Pi, though your Docker images do not support arm architectures. Nevertheless, it should not be so hard to integrate building images for arm devices in [your Build action](https://github.com/plausible/analytics/blob/master/.github/workflows/docker.yml).

Here are some useful links:

* [Running and building ARM Docker containers in x86
](https://ownyourbits.com/2018/06/27/running-and-building-arm-docker-containers-in-x86/)
* [docker/buildx: Building multi-platform images](https://github.com/docker/buildx#building-multi-platform-images)

Here are some examples in actions:

* **zigbee2mqtt** project builds its images like this: https://github.com/Koenkk/zigbee2mqtt/blob/master/.github/workflows/ci.yml#L58-L68
* **AdGuardHome** project:
  * GitHub Action: https://github.com/AdguardTeam/AdGuardHome/blob/master/.github/workflows/build.yml#L106-L111
  * A script for building: https://github.com/AdguardTeam/AdGuardHome/blob/master/scripts/make/build-docker.sh

There's even action for setting up a `buildx`: https://github.com/docker/setup-buildx-action

Please, consider that! This would open the possibility to host Plausible for lots of self-hosted and smart-home enthusiasts, like me. Thanks!

**Comment by ukutaht** [comment, marked as answer]
Clickhouse, our main database dependency, does not support ARM. So unfortunately we cannot support ARM either.

**Reply by ChiragSukhala** [reply]
ClickHouse has pre-production support for AArch64.
https://clickhouse.tech/#quick-start

So maybe explicitly mention this and include builds & instructions for ARM. 
The user will still need to install ClickHouse manually as Docker image is not available right now. (@ClickHouse if you are reading this, please include docker image builds for ARM)

**Comment by onthebridgetonowhere** [comment]
I wrote a [blog post](https://stefanstanciulescu.com/blog/plausible-analytics-on-raspberry-pi/) on how to self host using Docker on a Raspberry Pi 3. Hope this is useful to anyone who'd like to try to self host it.

**Reply by metmarkosaric** [reply]
nice, thanks for sharing @onthebridgetonowhere!

**Reply by Bennneeeet** [reply]
Hi @onthebridgetonowhere 
thank you for your how to.

As an Non-Expert it is some times a little bit hard to follow, but all in all great work.
I tried to set up Plausible on my Pi and started with the same situation as you: rpi 3 + fresh Ubuntu server installation.

I stucked two times at "Reverse proxy with caddy-gen". The first thing, I could resolve, is that in your snippet:
```
[...]
# Install Forego
RUN go get -u github.com/jwilder/forego &&
    cp /go/bin/forego /usr/bin/forego &&
    chmod u+x /usr/bin/forego
```
are missing the `\` for the new line(s) - that throws an error.

And the second part where I'm stucked is that I can't get build the caddy-gen container.
In the Dockerfile, the script stoped at step `11/19`. 

```
docker build -t caddy-gen .
Sending build context to Docker daemon  20.99kB
Step 1/19 : FROM caddy:2.4.3-alpine
 ---> 4c31b1758fe2
Step 2/19 : ARG DOCKER_GEN_VERSION="0.7.7"
 ---> Using cache
 ---> c804c3541ace
Step 3/19 : ARG FOREGO_VERSION="0.16.1"
 ---> Using cache
 ---> a9bf56094a74
Step 4/19 : ENV CADDYPATH="/etc/caddy"
 ---> Using cache
 ---> a669d5205cdb
Step 5/19 : ENV DOCKER_HOST="unix:///tmp/docker.sock"
 ---> Using cache
 ---> ab7e6cfe7285
Step 6/19 : RUN apk add --no-cache git make musl-dev go
 ---> Using cache
 ---> 4bf281676b5a
Step 7/19 : ENV GOROOT /usr/lib/go
 ---> Using cache
 ---> f0a3b850dff2
Step 8/19 : ENV GOPATH /go
 ---> Using cache
 ---> add9e56f3d86
Step 9/19 : ENV PATH /go/bin:$PATH
 ---> Using cache
 ---> 686b19ed080d
Step 10/19 : RUN  mkdir -p ${GOPATH}/src ${GOPATH}/bin
 ---> Using cache
 ---> c890fa58a1fc
Step 11/19 : RUN go get -u github.com/jwilder/forego && cp /go/bin/forego /usr/bin/forego &&  chmod u+x /usr/bin/forego
 ---> Running in b1f5432d946e
cannot find package "github.com/ddollar/forego/Godeps/_workspace/src/github.com/daviddengcn/go-colortext" in any of:
	/usr/lib/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/daviddengcn/go-colortext (from $GOROOT)
	/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/daviddengcn/go-colortext (from $GOPATH)
cannot find package "github.com/ddollar/forego/Godeps/_workspace/src/github.com/ddollar/dist" in any of:
	/usr/lib/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/ddollar/dist (from $GOROOT)
	/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/ddollar/dist (from $GOPATH)
cannot find package "github.com/ddollar/forego/Godeps/_workspace/src/github.com/kr/pretty" in any of:
	/usr/lib/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/kr/pretty (from $GOROOT)
	/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/kr/pretty (from $GOPATH)
cannot find package "github.com/ddollar/forego/Godeps/_workspace/src/github.com/subosito/gotenv" in any of:
	/usr/lib/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/subosito/gotenv (from $GOROOT)
	/go/src/github.com/ddollar/forego/Godeps/_workspace/src/github.com/subosito/gotenv (from $GOPATH)
The command '/bin/sh -c go get -u github.com/jwilder/forego && cp /go/bin/forego /usr/bin/forego &&  chmod u+x /usr/bin/forego' returned a non-zero code: 1

```
I don't know what the problem is. I tried different things, but here is some general knowledge missing. Can you tell me something about it or eventually the solution for it?

Thank you.

**Reply by onthebridgetonowhere** [reply]
@Bennneeeet  - the error you get seems to come from go. Using go version 1.13 I also get that error. Try to make sure go is up to date, e.g., v 1.17, and then it should not complain. Just to test, you could try add to the Dockerfile `RUN go version`, right after the line with `RUN apk add --no-cache git make musl-dev go`, and see what version you get. 

If you seem to have an older version (for some reason, the apk command does not get the latest Go version), you can try installing Go as following (add these in the Docker file, and remove #Install Go and # Configure Go related lines)
```
#we need the go version installed from apk to bootstrap the custom version built from source
RUN apk update && apk add go gcc bash musl-dev openssl-dev ca-certificates && update-ca-certificates

RUN wget https://golang.org/dl/go1.17.1.linux-arm64.tar.gz && tar -C /usr/local -xzf go1.17.1.linux-arm64.tar.gz

RUN cd /usr/local/go/src && ./make.bash

ENV PATH=$PATH:/usr/local/go/bin

#we delete the apk installed version to avoid conflict
RUN apk del go

RUN go version
```

The above steps I got from here [https://stackoverflow.com/questions/52056387/how-to-install-go-in-alpine-linux](https://stackoverflow.com/questions/52056387/how-to-install-go-in-alpine-linux)

Btw, you don't necessarily need reverse proxy with caddy-gen. It should work without it as well.

**Comment by paskal** [comment]
DockerHub image now doesn't work in arm64 with the error `standard_init_linux.go:228: exec user process caused: exec format error`.

I've tried to do the following to build the image myself:

```
git clone https://github.com/plausible/analytics.git && cd analytics && docker build --pull -t plausible/analytics:latest .
```

The image built that way works. I assume what is missing now is plausible themselves publishing builds for arm64 to DockerHub; that's it. I hope somebody will pick it up and enable automatic builds for arm64.

UPD: #2103 does precisely that.

**Reply by paskal** [reply]
Here is a docker-compose that works under arm64:

https://github.com/paskal/terrty/blob/d38e387/docker-compose.yml#L261-L334

**Reply by paskal** [reply]
UPD: it doesn't work, I think something is wrong either with clickhouse or its version (as I used a newer one), I will post an update here if I'll figure out how to make it work.

**Reply by paskal** [reply]
Starting with a clear DB on arm64 with docker-compose above yields infinite loading graphs on stats screen, and following errors. I think it might be possible due to using clickhouse `21.7.2.1` and not [recommended](https://github.com/plausible/hosting/blob/master/docker-compose.yml) `21.3.2.5`, will verify that on amd64 later.

<details>
  <summary>Logs</summary>


Plausible container:
```
13:12:27.604 request_id=FsmIEW3hvwZ12R0AAAkE [error] Handler {OpentelemetryPhoenix, :router_dispatch_exception} has failed and has been detached. Class=:error
Reason={:case_clause, []}
Stacktrace=[
  {:otel_telemetry, :peek_from_tracer_stack, 1,
   [file: 'src/otel_telemetry.erl', line: 94]},
  {:otel_telemetry, :set_current_telemetry_span, 2,
   [file: 'src/otel_telemetry.erl', line: 49]},
  {OpentelemetryPhoenix, :handle_router_dispatch_exception, 4,
   [file: 'lib/opentelemetry_phoenix.ex', line: 166]},
  {:telemetry, :"-execute/3-fun-0-", 4,
   [file: '/app/deps/telemetry/src/telemetry.erl', line: 150]},
  {:lists, :foreach, 2, [file: 'lists.erl', line: 1342]},
  {Phoenix.Router, :__call__, 2, [file: 'lib/phoenix/router.ex', line: 363]},
  {PlausibleWeb.Endpoint, :plug_builder_call, 2,
   [file: 'lib/plausible_web/endpoint.ex', line: 1]},
  {PlausibleWeb.Endpoint, :"call (overridable 3)", 2,
   [file: 'lib/plausible_web/endpoint.ex', line: 1]}
]

13:12:27.604 request_id=FsmIEW3hvo6OoXgAAAkh [error] Handler {OpentelemetryPhoenix, :router_dispatch_exception} has failed and has been detached. Class=:error
Reason={:case_clause, []}
Stacktrace=[
  {:otel_telemetry, :peek_from_tracer_stack, 1,
   [file: 'src/otel_telemetry.erl', line: 94]},
  {:otel_telemetry, :set_current_telemetry_span, 2,
   [file: 'src/otel_telemetry.erl', line: 49]},
  {OpentelemetryPhoenix, :handle_router_dispatch_exception, 4,
   [file: 'lib/opentelemetry_phoenix.ex', line: 166]},
  {:telemetry, :"-execute/3-fun-0-", 4,
   [file: '/app/deps/telemetry/src/telemetry.erl', line: 150]},
  {:lists, :foreach, 2, [file: 'lists.erl', line: 1342]},
  {Phoenix.Router, :__call__, 2, [file: 'lib/phoenix/router.ex', line: 363]},
  {PlausibleWeb.Endpoint, :plug_builder_call, 2,
   [file: 'lib/plausible_web/endpoint.ex', line: 1]},
  {PlausibleWeb.Endpoint, :"call (overridable 3)", 2,
   [file: 'lib/plausible_web/endpoint.ex', line: 1]}
]

13:12:27.604 [error] Task #PID<0.2891.0> started from #PID<0.2889.0> terminating
** (Clickhousex.Error) Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1)

    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:760: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:693: Ecto.Adapters.SQL.execute/5
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:219: Ecto.Repo.Queryable.execute/4
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:146: Ecto.Repo.Queryable.one/3
    (elixir 1.12.2) lib/task/supervised.ex:90: Task.Supervised.invoke_mfa/2
    (elixir 1.12.2) lib/task/supervised.ex:35: Task.Supervised.reply/5
    (stdlib 3.15) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Function: #Function<1.47805630/0 in Plausible.Stats.Aggregate.aggregate/3>
    Args: []
13:12:27.608 [error] Ranch protocol #PID<0.2889.0> of listener PlausibleWeb.Endpoint.HTTP (connection #PID<0.2888.0>, stream id 1) terminated
an exception was raised:
    ** (Clickhousex.Error) Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1)

        (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:760: Ecto.Adapters.SQL.raise_sql_call_error/1
        (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:693: Ecto.Adapters.SQL.execute/5
        (ecto 3.7.1) lib/ecto/repo/queryable.ex:219: Ecto.Repo.Queryable.execute/4
        (ecto 3.7.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
        (ecto 3.7.1) lib/ecto/repo/queryable.ex:146: Ecto.Repo.Queryable.one/3
        (elixir 1.12.2) lib/task/supervised.ex:90: Task.Supervised.invoke_mfa/2
        (elixir 1.12.2) lib/task/supervised.ex:35: Task.Supervised.reply/5
        (stdlib 3.15) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```
Plausible events DB, all traces are empty:
```
<Error> DynamicQueryHandler: Cannot send exception to client: Code: 24, e.displayText() = DB::Exception: Cannot write to ostream at offset 256, Stack trace (when copying this message, always include the lines below):
<Error> DynamicQueryHandler: Cannot send exception to client: Code: 24, e.displayText() = DB::Exception: Cannot write to ostream at offset 372, Stack trace (when copying this message, always include the lines below):
<Error> DynamicQueryHandler: Cannot send exception to client: Code: 24, e.displayText() = DB::Exception: Cannot write to ostream at offset 922, Stack trace (when copying this message, always include the lines below):
<Error> DynamicQueryHandler: Code: 24, e.displayText() = DB::Exception: Cannot write to ostream at offset 1264: While executing JSONCompactRowOutputFormat, Stack trace (when copying this message, always include the lines below):
<Error> DynamicQueryHandler: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream, Stack trace (when copying this message, always include the lines below):
<Error> DynamicQueryHandler: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream, Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 24, e.displayText() = DB::Exception: Cannot write to ostream at offset 1264: While executing JSONCompactRowOutputFormat (version 21.7.2.1) (from 172.19.0.8:48880) (in query: SELECT toDate(toTimeZone(s0."timestamp", 'Europe/Amsterdam')) as date FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) GROUP BY date ORDER BY date FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:48876) (in query: SELECT s0."screen_size", toUInt64(round(uniq(s0."user_id") * any(_sample_factor))) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) GROUP BY s0."screen_size" ORDER BY uniq(s0."user_id") DESC, min(s0."start") LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:48876) (in query: SELECT toUInt32(ifNotFinite(round(sum(is_bounce * sign) / sum(sign) * 100), 0)), toUInt32(ifNotFinite(round(avg(duration * sign)), 0)), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:48930) (in query: SELECT toUInt32(ifNotFinite(round(sum(is_bounce * sign) / sum(sign) * 100), 0)), toUInt32(ifNotFinite(round(avg(duration * sign)), 0)), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:48948) (in query: SELECT if(empty(s0."referrer_source"), 'Direct / None', s0."referrer_source"), toUInt64(round(uniq(s0."user_id") * any(_sample_factor))) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) AND (s0."referrer_source" != '') GROUP BY s0."referrer_source" ORDER BY uniq(s0."user_id") DESC, min(s0."start") LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:48954) (in query: SELECT s0."screen_size", toUInt64(round(uniq(s0."user_id") * any(_sample_factor))) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) GROUP BY s0."screen_size" ORDER BY uniq(s0."user_id") DESC, min(s0."start") LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:49076) (in query: SELECT s0."screen_size", toUInt64(round(uniq(s0."user_id") * any(_sample_factor))) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) GROUP BY s0."screen_size" ORDER BY uniq(s0."user_id") DESC, min(s0."start") LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `and(greaterOrEquals(start, '2021-12-12T23:00:00'), less(start, '2022-01-12T23:00:00'))` in source stream (version 21.7.2.1) (from 172.19.0.8:49092) (in query: SELECT toUInt32(ifNotFinite(round(sum(is_bounce * sign) / sum(sign) * 100), 0)), toUInt32(ifNotFinite(round(avg(duration * sign)), 0)), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "sessions" AS s0 sample 20000000 WHERE (s0."domain" = 'terrty.net') AND ((s0."start" >= '2021-12-12T23:00:00') AND (s0."start" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48930) (in query: SELECT toDate(toTimeZone(e0."timestamp", 'Europe/Amsterdam')) as date, toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY date ORDER BY date FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48948) (in query: SELECT e0."pathname", toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY e0."pathname" ORDER BY uniq(e0."user_id") DESC LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48950) (in query: SELECT toUInt64(round(uniq(e0."user_id") * any(_sample_factor))), toUInt64(round(countIf(e0."name" = 'pageview') * any(_sample_factor))), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48952) (in query: SELECT e0."pathname", toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY e0."pathname" ORDER BY uniq(e0."user_id") DESC LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48970) (in query: SELECT toDate(toTimeZone(e0."timestamp", 'Europe/Amsterdam')) as date, toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY date ORDER BY date FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:48972) (in query: SELECT toUInt64(round(uniq(e0."user_id") * any(_sample_factor))), toUInt64(round(countIf(e0."name" = 'pageview') * any(_sample_factor))), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:49070) (in query: SELECT e0."pathname", toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY e0."pathname" ORDER BY uniq(e0."user_id") DESC LIMIT 9 OFFSET 0 FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:49078) (in query: SELECT toDate(toTimeZone(e0."timestamp", 'Europe/Amsterdam')) as date, toUInt64(round(uniq(e0."user_id") * any(_sample_factor))) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) GROUP BY date ORDER BY date FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> executeQuery: Code: 8, e.displayText() = DB::Exception: Cannot find column `equals(domain, 'terrty.net')` in source stream (version 21.7.2.1) (from 172.19.0.8:49090) (in query: SELECT toUInt64(round(uniq(e0."user_id") * any(_sample_factor))), toUInt64(round(countIf(e0."name" = 'pageview') * any(_sample_factor))), if(any(_sample_factor) > 1, round(100 / any(_sample_factor)), 100) FROM "events" AS e0 sample 20000000 WHERE (e0."domain" = 'terrty.net') AND ((e0."timestamp" >= '2021-12-12T23:00:00') AND (e0."timestamp" < '2022-01-12T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
<Error> ServerErrorHandler: Poco::Exception. Code: 1000, e.code() = 32, e.displayText() = I/O error: Broken pipe, Stack trace (when copying this message, always include the lines below):
```
</details>

**Reply by paskal** [reply]
I didn't work with `lunalabsltd/clickhouse-server:21.7.2.7-arm`  but worked with `altinity/clickhouse-server:21.12.3.32.altinitydev.arm` as suggested below by @ukutaht.

**Reply by ukutaht** [reply]
great!

**Reply by paskal** [reply]
Everything works with `clickhouse/clickhouse-server:22.6-alpine` smoothly, the documentation is already updated to use it in https://github.com/plausible/hosting/pull/45.

**Comment by clement-escolano** [comment]
ClickHouse has now official ARM64 Docker images (related [issue](https://github.com/ClickHouse/ClickHouse/issues/22222)). Is there anything else blocking the arm64 support of plausible?

**Reply by ukutaht** [reply]
I am not aware of any other blockers

**Comment by ukutaht** [comment]
Anyone interested in testing on ARM, you can try running our standard setup with the clickhouse image swapped for [this official ARM one](https://hub.docker.com/layers/altinity/clickhouse-server/21.12.3.32.altinitydev.arm/images/sha256-700e2734bb8b1a1eefe3660b8e5284eabd1c8e2adaf3de4180d87af21c5f470c?context=explore). Curious to hear if it works with Plausible.

**Reply by mve** [reply]
Just tested it on m1 pro, seems to work for me.
build plausible/analytics image on the system and used that image in the docker-compose. Updated clickhouse to the image you linked.

docker-compose looks like this:

``` yml 

version: "3.3"
services:
  mail:
    image: bytemark/smtp
    restart: always

  plausible_db:
    image: postgres:12
    restart: always
    volumes:
      - db-data:/var/lib/postgresql/data
    environment:
      - POSTGRES_PASSWORD=postgres

  plausible_events_db:
    image: altinity/clickhouse-server:21.12.3.32.altinitydev.arm
    restart: always
    volumes:
      - event-data:/var/lib/clickhouse
      - ./clickhouse/clickhouse-config.xml:/etc/clickhouse-server/config.d/logging.xml:ro
      - ./clickhouse/clickhouse-user-config.xml:/etc/clickhouse-server/users.d/logging.xml:ro
    ulimits:
      nofile:
        soft: 262144
        hard: 262144

  plausible:
    build:
      context: ../plausible-analytics-source
      dockerfile: Dockerfile
    restart: always
    command: sh -c "sleep 10 && /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh db init-admin && /entrypoint.sh run"
    depends_on:
      - plausible_db
      - plausible_events_db
      - mail
    ports:
      - 8000:8000
    env_file:
      - plausible-conf.env

volumes:
  db-data:
    driver: local
  event-data:
    driver: local
  geoip:
    driver: local
```

"plausible-analytics-source" is just the master branch of this repo.

I was able to run the container, log in, add a site from localhost and have data show up in the dashboard.
It does give me this warning, not sure if this is relevant:

``` 
plausible_events_db_1  | libunwind: Unsupported .eh_frame_hdr version
```

bytemark/smtp was running emulated, but it was able to send an email to me no problem.
![image](https://user-images.githubusercontent.com/38584349/150579899-431ed5c8-c35d-43f3-9a87-4df223880049.png)

Let me know if you need me to check anything specific but it looks promising from what I can see.

**Reply by ukutaht** [reply]
Thanks, that's great! I'll feed that warning back to the clickhouse team so they are aware of it as well. Once they release a stable image I think we should add instructions to the hosting repo for this.

**Comment by clement-escolano** [comment]
Official ARM image for ClickHouse is now available 🍾 
Issue link: https://github.com/ClickHouse/ClickHouse/issues/22222
Docker hub: https://hub.docker.com/r/clickhouse/clickhouse-server/tags

**Comment by navanchauhan** [comment]
Tested clickhouse/clickhouse-server:22.4.2.1 on x86_64 and everything appears to work! For arm64, I am getting an exec error, but that is because plausible does not publish arm64 images (and the error is not from clickhouse).

Maybe it is time to apply the patch(es) from https://github.com/plausible/analytics/pull/1287?

Edit: Testing by using arm64 version of the dependencies, everything works on an arm64 machine. The mentioned PR patch should definitely help

**Reply by just5ky** [reply]
@navanchauhan can you try this image `docker pull justsky/analytics`
I build it for x86 and arm64

**Reply by navanchauhan** [reply]
it is erroring out because of the mail image `mail_1                 | standard_init_linux.go:228: exec user process caused: exec format error`

let me see if I can rebuild that for arm64 or disable it

**Reply by just5ky** [reply]
if you are using `bytemark/smtp` it will error out, it's only built for amd64.

**Reply by navanchauhan** [reply]
Just built and published `navanchauhan/smtp` which should support arm64

**Reply by navanchauhan** [reply]
It works! In https://github.com/plausible/hosting just replace:
* `bytemark/smtp` with `navanchauhan/smtp` 
* `plausible/analytics` with `justsky/analytics`
* `yandex/clickhouse-server:21.3.2.5` with `clickhouse/clickhouse-server:22.4.2.1`

I haven't thoroughly checked the difference between using different versions of clickhouse, but I am able to login to the admin dashboard in plausible

**Reply by just5ky** [reply]
you can use the latest tag for clickhouse-sever

**Reply by paskal** [reply]
To mitigate https://github.com/ClickHouse/ClickHouse/issues/15638, we'll need to add file like `/etc/clickhouse-server/config.d/clickhouse-issue-15638.xml` to `plausible_events_db` with the following content:
```xml
<clickhouse>
    <profiles>
        <default>
            <query_profiler_real_time_period_ns>0</query_profiler_real_time_period_ns>
            <query_profiler_cpu_time_period_ns>0</query_profiler_cpu_time_period_ns>
        </default>
    </profiles>
</clickhouse>
```

Otherwise, clickhouse running on ARM spams logs with `libunwind: Unsupported .eh_frame_hdr version`.

I've prepared https://github.com/plausible/hosting/pull/45 which would make plausible/hosting more friendly to ARM out of the box.

**Reply by paskal** [reply]
UPD: no need for a workaround as the issue is [fixed](https://github.com/ClickHouse/ClickHouse/pull/37797) in the clickhouse-server master and will eventually be released in the next version.

**Comment by ruslandoga** [comment]
👋

[`v1.5.0-rc.0`](https://github.com/plausible/analytics/discussions/2324) is out with arm64 support! Please try it out :)

**Reply by paskal** [reply]
Seem to work for me the same as self-build master was working for the last few months, everything is good.

**Reply by HelixHEX** [reply]
Thanks, this works perfect on my first gen M1 mac!
